File: backend/services/scan_trade_hint.py
"""Buy vs wait conviction for scan candidates (research hint — not a trade order)."""
from __future__ import annotations

from typing import Any

from engines.recommendation.engine import recommendation_label_from_score
from models.schemas import RiskLevel

_RISK_WAIT_BONUS: dict[RiskLevel, float] = {
    RiskLevel.low: 0.0,
    RiskLevel.medium: 12.0,
    RiskLevel.high: 28.0,
}


def _normalize_pair(buy_raw: float, wait_raw: float) -> tuple[float, float]:
    buy_raw = max(0.0, buy_raw)
    wait_raw = max(0.0, wait_raw)
    total = buy_raw + wait_raw
    if total <= 0:
        return 50.0, 50.0
    buy_pct = round(100.0 * buy_raw / total, 1)
    wait_pct = round(100.0 - buy_pct, 1)
    return buy_pct, wait_pct


def _gate_label(
    label: str,
    *,
    data_quality_score: float | None,
    provider_limited: bool,
    earnings_soon: bool,
    risk_level: RiskLevel,
) -> tuple[str, list[str]]:
    gates: list[str] = []
    if data_quality_score is not None and data_quality_score < 35:
        return "high_risk_no_decision", ["Insufficient data quality"]
    if provider_limited:
        if label in ("strong_buy", "buy"):
            label = "watch"
        gates.append("Partial provider data")
    if data_quality_score is not None and data_quality_score < 50:
        if label in ("strong_buy", "buy"):
            label = "watch"
        gates.append("Low data quality")
    if earnings_soon and label == "strong_buy":
        label = "buy"
        gates.append("Earnings soon")
    if risk_level == RiskLevel.high and label == "strong_buy":
        label = "buy"
        gates.append("Elevated risk")
    return label, gates


def _short_reason(
    label: str,
    *,
    score: float,
    gates: list[str],
) -> str:
    if gates:
        return gates[0]
    if label in ("strong_buy", "buy"):
        return f"Score {score:.0f} supports entry"
    if label == "watch":
        return "Setup forming — confirm before sizing"
    if label == "avoid":
        return "Weak score — skip for now"
    if label == "high_risk_no_decision":
        return "Data too thin for a confident call"
    return "Mixed signals — patience"


def compute_scan_trade_hint(
    *,
    score: float,
    sleeve: str,
    risk_level: RiskLevel,
    data_quality_score: float | None = None,
    earnings_soon: bool = False,
    provider_limited: bool = False,
) -> dict[str, Any]:
    """Return recommendation label plus buy/wait mix for a scan row."""
    score = max(0.0, min(100.0, float(score)))

    buy_raw = max(0.0, (score - 35.0) * 1.4)
    wait_raw = max(5.0, 100.0 - score * 0.85) + _RISK_WAIT_BONUS.get(risk_level, 12.0)

    if earnings_soon:
        wait_raw += 15.0
        buy_raw *= 0.75
    if provider_limited:
        wait_raw += 20.0
        buy_raw *= 0.6

    if data_quality_score is not None:
        dq = float(data_quality_score)
        if dq < 50:
            buy_raw *= 0.55
            wait_raw += 18.0
        if dq < 35:
            buy_raw *= 0.3
            wait_raw += 25.0

    if sleeve == "penny":
        if score < 58:
            buy_raw *= 0.5
            wait_raw += 10.0
        if risk_level == RiskLevel.high:
            buy_raw *= 0.65
    elif sleeve == "compounder":
        if score < 62:
            buy_raw *= 0.45
            wait_raw += 8.0

    buy_pct, wait_pct = _normalize_pair(buy_raw, wait_raw)
    label = recommendation_label_from_score(score)
    label, gates = _gate_label(
        label,
        data_quality_score=data_quality_score,
        provider_limited=provider_limited,
        earnings_soon=earnings_soon,
        risk_level=risk_level,
    )

    return {
        "recommendation": label,
        "buy_pct": buy_pct,
        "wait_pct": wait_pct,
        "trade_hint_reason": _short_reason(label, score=score, gates=gates),
    }


def attach_trade_hint_to_metrics(
    metrics: dict[str, Any],
    *,
    score: float,
    sleeve: str,
    risk_level: RiskLevel,
    data_quality_score: float | None = None,
    earnings_soon: bool = False,
    provider_limited: bool = False,
) -> dict[str, Any]:
    out = dict(metrics)
    out.update(
        compute_scan_trade_hint(
            score=score,
            sleeve=sleeve,
            risk_level=risk_level,
            data_quality_score=data_quality_score,
            earnings_soon=earnings_soon or bool(metrics.get("earnings_soon")),
            provider_limited=provider_limited or bool(metrics.get("provider_limited_partial_data")),
        )
    )
    return out

File: frontend/src/lib/scanTradeHint.ts
import type { StockResult } from "@/lib/types";

export type ScanTradeHint = {
  recommendation: string;
  buyPct: number;
  waitPct: number;
  reason: string;
  stability?: { score: number; classification: string };
  entryRisk?: { score: number; classification: string; source?: string };
};

function decisionLayers(metrics: Record<string, unknown>): Pick<ScanTradeHint, "stability" | "entryRisk"> {
  const stabilityScore = metrics.stability_score;
  const stabilityClassification = metrics.stability_classification;
  const entryRiskScore = metrics.entry_risk_score;
  const entryRiskClassification = metrics.entry_risk_classification;
  const entryRiskSource = metrics.entry_risk_source;
  return {
    stability:
      typeof stabilityScore === "number" && typeof stabilityClassification === "string"
        ? { score: stabilityScore, classification: stabilityClassification }
        : undefined,
    entryRisk:
      typeof entryRiskScore === "number" && typeof entryRiskClassification === "string"
        ? {
            score: entryRiskScore,
            classification: entryRiskClassification,
            source: typeof entryRiskSource === "string" ? entryRiskSource : undefined,
          }
        : undefined,
  };
}

function riskWaitBonus(level: string): number {
  if (level === "low") return 0;
  if (level === "high") return 28;
  return 12;
}

function labelFromScore(score: number): string {
  if (score >= 80) return "strong_buy";
  if (score >= 65) return "buy";
  if (score >= 50) return "watch";
  if (score >= 35) return "hold";
  return "avoid";
}

/** Client fallback when cached scan rows predate trade-hint metrics. */
export function deriveScanTradeHint(stock: StockResult): ScanTradeHint {
  const score = Math.max(0, Math.min(100, stock.score));
  const sleeve = stock.bucket;
  const risk = stock.risk_level;
  const metrics = stock.metrics ?? {};
  const dq = metrics.data_quality_score as number | undefined;
  const earningsSoon = Boolean(stock.earnings_soon ?? metrics.earnings_soon);
  const providerLimited = Boolean(metrics.provider_limited_partial_data);
  const fallbackReason =
    typeof metrics.fallback_reason === "string" ? metrics.fallback_reason : "";
  const nonProviderFallback =
    Boolean(fallbackReason) &&
    fallbackReason !== "none" &&
    !providerLimited;

  let buyRaw = Math.max(0, (score - 35) * 1.4);
  let waitRaw = Math.max(5, 100 - score * 0.85) + riskWaitBonus(risk);

  if (earningsSoon) {
    waitRaw += 15;
    buyRaw *= 0.75;
  }
  if (providerLimited) {
    waitRaw += 20;
    buyRaw *= 0.6;
  } else if (nonProviderFallback) {
    waitRaw += 12;
    buyRaw *= 0.75;
  }
  if (dq != null) {
    if (dq < 50) {
      buyRaw *= 0.55;
      waitRaw += 18;
    }
    if (dq < 35) {
      buyRaw *= 0.3;
      waitRaw += 25;
    }
  }
  if (sleeve === "penny") {
    if (score < 58) {
      buyRaw *= 0.5;
      waitRaw += 10;
    }
    if (risk === "high") buyRaw *= 0.65;
    if (metrics.stability_classification === "rejected") {
      buyRaw *= 0.2;
      waitRaw += 40;
    }
    if (["wait", "extended", "no_chase"].includes(String(metrics.entry_risk_classification ?? ""))) {
      buyRaw *= 0.35;
      waitRaw += 35;
    }
  } else if (sleeve === "compounder" && score < 62) {
    buyRaw *= 0.45;
    waitRaw += 8;
  }

  const total = buyRaw + waitRaw || 1;
  const buyPct = Math.round((100 * buyRaw) / total * 10) / 10;
  const waitPct = Math.round((100 - buyPct) * 10) / 10;

  let recommendation = labelFromScore(score);
  let reason = "Mixed signals — patience";
  if (dq != null && dq < 35) {
    recommendation = "high_risk_no_decision";
    reason = "Insufficient data quality";
  } else if (providerLimited && (recommendation === "strong_buy" || recommendation === "buy")) {
    recommendation = "watch";
    reason = "Partial provider data";
  } else if (
    nonProviderFallback &&
    (recommendation === "strong_buy" || recommendation === "buy")
  ) {
    recommendation = "watch";
    reason = "Fallback ranking — verify filters";
  } else if (dq != null && dq < 50 && (recommendation === "strong_buy" || recommendation === "buy")) {
    recommendation = "watch";
    reason = "Low data quality";
  } else if (recommendation === "strong_buy" || recommendation === "buy") {
    reason = `Score ${score.toFixed(0)} supports entry`;
  } else if (recommendation === "watch") {
    reason = "Setup forming — confirm before sizing";
  } else if (recommendation === "avoid") {
    reason = "Weak score — skip for now";
  }

  if (sleeve === "penny" && metrics.stability_classification === "rejected") {
    recommendation = "avoid";
    reason = "Strong stock or not, stability gate failed";
  } else if (
    sleeve === "penny" &&
    ["wait", "extended", "no_chase"].includes(String(metrics.entry_risk_classification ?? ""))
  ) {
    recommendation = "watch";
    reason = `Good candidate, poor entry (${String(metrics.entry_risk_classification).replaceAll("_", " ")})`;
  }

  return { recommendation, buyPct, waitPct, reason, ...decisionLayers(metrics) };
}

export function getScanTradeHint(stock: StockResult): ScanTradeHint {
  const metrics = stock.metrics ?? {};
  const buyPct = metrics.buy_pct as number | undefined;
  const waitPct = metrics.wait_pct as number | undefined;
  const recommendation = metrics.recommendation as string | undefined;
  const reason = metrics.trade_hint_reason as string | undefined;

  if (
    typeof buyPct === "number" &&
    typeof waitPct === "number" &&
    recommendation &&
    typeof reason === "string"
  ) {
    return { recommendation, buyPct, waitPct, reason, ...decisionLayers(metrics) };
  }
  return deriveScanTradeHint(stock);
}
